scapy/contrib/automotive/xcp/xcp.py

```diff
@@ -0,0 +1,462 @@
+    ]
+
+
+for pid in range(0, 0xBF + 1):
+    bind_layers(CTORequest, DTO, pid=pid)
+
+
+class CTOResponse(Packet):
+    packet_codes = {
+        0xFF: "RES",
+        0xFE: "ERR",
+        0xFD: "EV",
+        0xFC: "SERV",
+    }
+    name = "Command Transfer Object Response"
+
+    fields_desc = [
+        ByteEnumField("packet_code", 0, packet_codes),
+    ]
+
+    @staticmethod
+    def get_positive_response_cls(request):
+        # The pid of the request this packet is the response for
+        request_pid = request.pid
+        # First check the special cases with sub commands
+        # They can't be fit in a simple dictionary,
+        # so deal with them separately
+        if request_pid == 0xF2:
+            if request.sub_command_code == 255:
+                return TransportLayerCmdGetSlaveIdResponse
+            if request.sub_command_code == 254:
+                return TransportLayerCmdGetDAQIdResponse
+        if request_pid == 0xE8:
+            if request.mode == "get_basic_address_info":
+                return SegmentInfoMode0PositiveResponse
+            if request.mode == "get_standard_info":
+                return SegmentInfoMode1PositiveResponse
+            if request.mode == "get_address_mapping_info":
+                return SegmentInfoMode2PositiveResponse
+        return {0xFF: ConnectPositiveResponse,
+                0xFD: StatusPositiveResponse,
+                0xFB: CommonModeInfoPositiveResponse,
+                0xFA: IdPositiveResponse,
+                0xF8: SeedPositiveResponse,
+                0xF7: UnlockPositiveResponse,
+                0xF5: UploadPositiveResponse,
+                0xF4: ShortUploadPositiveResponse,
+                0xF3: ChecksumPositiveResponse,
+                0xEA: CalPagePositiveResponse,
+                0xE9: PagProcessorInfoPositiveResponse,
+                0xE7: PageInfoPositiveResponse,
+                0xE5: SegmentModePositiveResponse,
+                0xDF: DAQListModePositiveResponse,
+                0xDE: StartStopDAQListPositiveResponse,
+                0xDC: DAQClockListPositiveResponse,
+                0xDB: ReadDAQPositiveResponse,
+                0xDA: DAQProcessorInfoPositiveResponse,
+                0xD9: DAQResolutionInfoPositiveResponse,
+                0xD8: DAQListInfoPositiveResponse,
+                0xD7: DAQEventInfoPositiveResponse,
+                0xD2: ProgramStartPositiveResponse,
+                0xCE: PgmProcessorPositiveResponse,
+                0xCD: SectorInfoPositiveResponse,
+                }.get(request_pid, GenericResponse)
+
+    def answers(self, request):
+        """In XCP, the payload of a response packet is dependent on the pid
+        field of the corresponding request.
+        This method changes the class of the payload to the class
+        which is expected for the given request."""
+        if not isinstance(request, CTORequest) or \
+                not isinstance(self, CTOResponse):
```

Comment:
can this condition be False?

Reply:
I don'tn know about `if not isinstance(request, CTORequest)` but `not isinstance(self, CTOResponse)`should never be false..